spec/support/crud.rb

```diff
@@ -130,25 +130,29 @@ def initialize(data, test)
         end
       end
 
-      def expected_outcome
-        @operations.last.outcome
-      end
+      # Operations to be performed by the test.
+      #
+      # For CRUD tests, there is one operation for test. For retryable writes,
+      # there are multiple operations for each test. In either case we build
+      # an array of operations.
+      attr_reader :operations
 
       # Run the test.
       #
-      # @example Run the test.
-      #   test.run(collection)
+      # The specified number of operations are executed, so that the
+      # test can assert on the outcome of each specified operation in turn.
       #
       # @param [ Collection ] collection The collection the test
       #   should be run on.
+      # @param [ Integer ] num_ops Number of operations to run.
       #
       # @return [ Result, Array<Hash> ] The result(s) of running the test.
       #
       # @since 2.0.0
-      def run(collection)
+      def run(collection, num_ops)
         result = nil
-        @operations.each do |op|
-          result = op.execute(collection)
+        1.upto(num_ops) do |i|
```

Comment:
This feels like a really strange way to implement this; instead of just running each operation once and checking the result after each one, we're instead running the first operation, then the first two (including the one we just did), then the first three (including the two we just did), etc. I feel like refactoring this a bit more rather than calling run N times is worth it to avoid running twice as many operations.

Reply:
This is done intentionally to make it possible to run each example individually. SDAM tests currently run each operation once and this makes them hard to debug (because one has to skip over assertions for each operation) and awkward to parallelize (if the entire file is not run sequentially by the same process the run would fail).

Reply:
Fair enough